File: app/src/main/java/com/raival/fileexplorer/tabs/file/tasks/CutTask.java
package com.raival.fileexplorer.tabs.file.tasks;

import android.os.Handler;
import android.os.Looper;

import com.raival.fileexplorer.tabs.file.model.Task;
import com.raival.fileexplorer.utils.FileUtil;

import java.io.File;
import java.util.ArrayList;

public class CutTask extends Task {
    private final ArrayList<File> filesToCut;
    private File activeDirectory;

    public CutTask(ArrayList<File> selectedFiles) {
        filesToCut = selectedFiles;
    }

    public ArrayList<File> getFilesToCut() {
        return filesToCut;
    }

    @Override
    public String getName() {
        return "Cut";
    }

    @Override
    public String getDetails() {
        StringBuilder sb = new StringBuilder();
        boolean first = true;

        for (File file : filesToCut) {
            if (!first) {
                sb.append(", ");
            }
            sb.append(file.getName());
            first = false;
        }
        return sb.toString();
    }

    @Override
    public boolean isValid() {
        for (File file : filesToCut) {
            if (!file.exists())
                return false;
        }
        return true;
    }

    @Override
    public void setActiveDirectory(File directory) {
        activeDirectory = directory;
    }

    @Override
    public void start(OnUpdateListener onUpdateListener, OnFinishListener onFinishListener) {
        new Thread(() -> {
            boolean error = false;
            try {
                int progress = 1;
                for(File file : filesToCut){
                    try{
                        int finalProgress = progress;
                        new Handler(Looper.getMainLooper()).post(() -> {
                            onUpdateListener.onUpdate("["
                                    + finalProgress
                                    + "/"
                                    + filesToCut.size()
                                    + "]"
                                    + "Moving "
                                    + file.getName());
                        });
                        FileUtil.move(file, activeDirectory);
                    } catch (Exception exception){
                        error = true;
                    }
                    ++progress;
                }
                boolean finalError = error;
                new Handler(Looper.getMainLooper()).post(() -> onFinishListener.onFinish(finalError
                        ? "An error occurred, some files haven't been moved"
                        : "Files have been moved successfully"));
            } catch (Exception e) {
                e.printStackTrace();
                new Handler(Looper.getMainLooper()).post(() -> onFinishListener.onFinish(e.toString()));
            }
        }).start();
    }
}

File: app/src/main/java/com/raival/fileexplorer/tab/file/task/CutTask.kt
package com.raival.fileexplorer.tab.file.task

import com.raival.fileexplorer.tab.file.misc.FileUtils
import com.raival.fileexplorer.tab.file.model.Task
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File

class CutTask(private val filesToCut: ArrayList<File>) : Task() {
    private var activeDirectory: File? = null
    override val name: String
        get() = "Cut"
    override val details: String
        get() {
            val sb = StringBuilder()
            var first = true
            for (file in filesToCut) {
                if (!first) {
                    sb.append(", ")
                }
                sb.append(file.name)
                first = false
            }
            return sb.toString()
        }
    override val isValid: Boolean
        get() {
            for (file in filesToCut) {
                if (!file.exists()) return false
            }
            return true
        }

    override fun setActiveDirectory(directory: File) {
        activeDirectory = directory
    }

    override fun start(onUpdate: OnUpdateListener, onFinish: OnFinishListener) {
        CoroutineScope(Dispatchers.IO).launch {
            var error = false
            try {
                var progress = 1
                for (file in filesToCut) {
                    try {
                        val finalProgress = progress
                        withContext(Dispatchers.Main) {
                            onUpdate.onUpdate(
                                "["
                                        + finalProgress
                                        + "/"
                                        + filesToCut.size
                                        + "]"
                                        + "Moving "
                                        + file.name
                            )
                        }
                        FileUtils.move(file, activeDirectory)
                    } catch (exception: Exception) {
                        error = true
                    }
                    ++progress
                }
                val finalError = error
                withContext(Dispatchers.Main) { onFinish.onFinish(if (finalError) "An error occurred, some files haven't been moved" else "Files have been moved successfully") }
            } catch (e: Exception) {
                e.printStackTrace()
                withContext(Dispatchers.Main) { onFinish.onFinish(e.toString()) }
            }
        }
    }
}